### DISCUSSION TITLE
Undefined is not a function in expo app

### DISCUSSION THREAD

**Discussion by elisalimli** [urql-graphql/urql#3203] [Q&A] [answered]
### When I try to write a simple screen to get `hello` query and print it to the screen the `Undefined is not a function` error occurs. 

CODE :

```tsx
import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { use query, gql } from "urql";

const HELLO_QUERY = gql`
  query {
    hello
  }
`;

export default function HelloScreen() {
  const [{ data, fetching, error }] = useQuery({ query: HELLO_QUERY });

  if (fetching) {
    return <Text>Loading...</Text>;
  }

  if (error) {
    return <Text>Error: {error.message}</Text>;
  }

  return (
    <View style={styles.container}>
      <Text style={styles.title}>Hello</Text>
      <Text style={styles.text}>{data.hello}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#fff",
  },
  title: {
    fontSize: 32,
    marginBottom: 16,
  },
  text: {
    fontSize: 24,
  },
});

```

-------------------

I resolved this error by downgrading the Urql version from 4.0.2 to 3.0.4.

What causes this problem and how to solve it? 

### Urql version

urql v4.0.2

### Validations

- [X] I can confirm that this is a bug report, and not a feature request, RFC, question, or discussion, for which [GitHub Discussions](https://github.com/urql-graphql/urql/discussions) should be used
- [X] Read the [docs](https://formidable.com/open-source/urql/docs/).
- [X] Follow our [Code of Conduct](https://github.com/urql-graphql/urql/blob/main/CODE_OF_CONDUCT.md)

**Comment by kitten** [comment]
I'm going to he honest here, there's a high likelihood that you didn't check the migration guide for v4: https://github.com/urql-graphql/urql/issues/3114

There's likely one of two problems in your app:
- You haven't created a client and provider
- You haven't passed the default exchanges to your client

However, without a full error message with a stack trace or a full description, I can't really tell.

Just to point this out, we've got Discord GitHub Discussions to help with debugging and questions, because it's easier to have a back and forth there 👀

**Reply by elisalimli** [reply]
Is there anything I missed?

**Comment by elisalimli** [comment]


**Reply by kitten** [reply]
Converted to a discussion. 😅 Again though, I'm not a human linter, without a full error message and a stack trace, I can't really help you

**Comment by elisalimli** [comment]
Thanks for letting me know. Here is the full error message and stack trace:

```
 ERROR  TypeError: undefined is not a function

This error is located at:
    in TabTwoScreen
    in Unknown (created by Route(two))
    in Route (created by Route(two))
    in LocationProvider (created by Route(two))
    in Route(two) (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by BottomTabView)
    in RCTView (created by View)
    in View (created by Screen)
    in RCTView (created by View)
    in View (created by Background)
    in Background (created by Screen)
    in Screen (created by BottomTabView)
    in RNSScreen (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by InnerScreen)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze (created by InnerScreen)
    in InnerScreen (created by Screen)
    in Screen (created by MaybeScreen)
    in MaybeScreen (created by BottomTabView)
    in RNSScreenNavigationContainer (created by ScreenContainer)
    in ScreenContainer (created by MaybeScreenContainer)
    in MaybeScreenContainer (created by BottomTabView)
    in RCTView (created by View)
    in View (created by SafeAreaInsetsContext)
    in SafeAreaProviderCompat (created by BottomTabView)
    in BottomTabView (created by BottomTabNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by BottomTabNavigator)
    in BottomTabNavigator
    in Unknown (created by TabLayout)
    in TabLayout
    in Unknown (created by Route((tabs)))
    in Route (created by Route((tabs)))
    in LocationProvider (created by Route((tabs)))
    in Route((tabs)) (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by SceneView)
    in RCTView (created by View)
    in View (created by DebugContainer)
    in DebugContainer (created by MaybeNestedStack)
    in MaybeNestedStack (created by SceneView)
    in RCTView (created by View)
    in View (created by SceneView)
    in RNSScreen (created by AnimatedComponent)
    in AnimatedComponent
    in AnimatedComponentWrapper (created by InnerScreen)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze (created by InnerScreen)
    in InnerScreen (created by Screen)
    in Screen (created by SceneView)
    in SceneView (created by NativeStackViewInner)
    in Suspender (created by Freeze)
    in Suspense (created by Freeze)
    in Freeze (created by DelayedFreeze)
    in DelayedFreeze (created by ScreenStack)
    in RNSScreenStack (created by ScreenStack)
    in ScreenStack (created by NativeStackViewInner)
    in NativeStackViewInner (created by NativeStackView)
    in RCTView (created by View)
    in View (created by SafeAreaInsetsContext)
    in SafeAreaProviderCompat (created by NativeStackView)
    in NativeStackView (created by NativeStackNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by NativeStackNavigator)
    in NativeStackNavigator
    in Unknown (created by RootLayoutNav)
    in ThemeProvider (created by RootLayoutNav)
    in RootLayoutNav (created by RootLayout)
    in RootLayout
    in Try
    in Unknown
    in Unknown (created by Route())
    in Route (created by Route())
    in LocationProvider (created by Route())
    in Route() (created by RootRoute)
    in RootRoute (created by ContextNavigator)
    in InitialRootStateProvider (created by ContextNavigator)
    in EnsureSingleNavigator
    in BaseNavigationContainer
    in ThemeProvider
    in NavigationContainerInner (created by NavigationContainer)
    in NavigationContainer (created by ContextNavigator)
    in RootRouteNodeProvider (created by ContextNavigator)
    in ContextNavigator (created by ExpoRoot)
    in RNCSafeAreaProvider (created by SafeAreaProvider)
    in SafeAreaProvider (created by ExpoRoot)
    in ExpoRoot (created by App)
    in App (created by withDevTools(App))
    in withDevTools(App) (created by ErrorOverlay)
    in ErrorToastContainer (created by ErrorOverlay)
    in ErrorOverlay
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent), js engine: hermes

```

-------------
And this is my TabTwoScreen:

```tsx 
import { StyleSheet } from "react-native";

import EditScreenInfo from "../../components/EditScreenInfo";
import { Text, View } from "../../components/Themed";
import React from "react";
import { useQuery } from "urql";
import { meQueryDocument } from "../../src/graphql/query/user/me";

export default function TabTwoScreen() {
  const [result] = useQuery({ query: meQueryDocument });
  console.log("me query result", result);
  if (result?.fetching) {
    return <Text>Loading...</Text>;
  }

  return (
    <View style={styles.container}>
      <Text style={styles.title}>Tab Two</Text>
      <Text style={styles.title}>Hello, {result?.data?.me?.username}</Text>
      <View
        style={styles.separator}
        lightColor="#eee"
        darkColor="rgba(255,255,255,0.1)"
      />
      <EditScreenInfo path="app/(tabs)/two.tsx" />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
  },
  separator: {
    marginVertical: 30,
    height: 1,
    width: "80%",
  },
});

```

**Reply by kitten** [reply]
> in main(RootComponent), js engine: hermes

I think this may relate to an issue with Expo/React Native not activating the async generator / async iterable polyfill properly: https://github.com/urql-graphql/urql/issues/3129

**Reply by elisalimli** [reply]
Added missing `Symbol.asyncIterator` in hermes as said [#10764](https://github.com/aws-amplify/amplify-js/issues/10764#issuecomment-1363140556).

Now I'm getting a network error :

`[Network] Cannot convert undefined value to object `



`
    at construct (native)
    at apply (native)
    at _construct (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:5474:28)
    at Wrapper (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:5436:25)
    at construct (native)
    at _createSuperInternal (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:131033:294)
    at call (native)
    at CombinedError (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:131069:26)
    at makeErrorResult (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:131329:31)
    at ?anon_0_ (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:131653:30)
    at throw (native)
    at resume (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:22509:30)
    at anonymous (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:22520:17)
    at tryCallOne (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:53:16)
    at anonymous (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:139:27)
    at apply (native)
    at anonymous (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:49786:26)
    at _callTimer (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:49705:17)
    at _callReactNativeMicrotasksPass (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:49735:17)
    at callReactNativeMicrotasks (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:49898:44)
    at __callReactNativeMicrotasks (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:27315:46)
    at anonymous (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:27127:45)
    at __guard (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:27299:15)
    at flushedQueue (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:27126:21)
    at callFunctionReturnFlushedQueue (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:27111:33)
`

**Reply by kitten** [reply, marked as answer]
I haven't really had time to dig into this and figure out why the polyfills and/or transpilations are being applied, but I'm pretty sure you can also for now force your app back to JSC from Hermes instead, if you don't mind that 🤔

Edit: (link) https://docs.expo.dev/guides/configuring-js-engines/

**Reply by elisalimli** [reply]
I also tried to switch back to JSC. But it didn't solve the issue.

Thanks for your help.

**Reply by dariocravero** [reply]
Wanted to chime in here to first of all thank you both for the steps to get to this.

I can confirm that importing async-iterator as suggested here https://github.com/aws-amplify/amplify-js/issues/10764#issuecomment-1363140556 works for us using Expo v48 with Hermes. We added it to `index.js` at the very top like this:

```
import 'core-js/full/symbol/async-iterator'
import { registerRootComponent } from 'expo'

import App from './App.js'

// registerRootComponent calls AppRegistry.registerComponent('main', () => App);
// It also ensures that whether you load the app in Expo Go or in a native build,
// the environment is set up appropriately
registerRootComponent(App)
```

**Reply by elisalimli** [reply]
@dariocravero 

You're welcome! In my case, today I re-tried adding the pollyfill stuff to index.ts.Although that fixes the 'undefined is not a function' error, but again the same network error occurs. Do you use new expo-router with urql ? 
> `[Network] Cannot convert undefined value to object at construct (native) at apply (native) at _construct (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:7125:28) at Wrapper (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:7087:25) at construct (native) at _createSuperInternal (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:128706:294) at call (native) at CombinedError (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:128742:26) at makeErrorResult (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:129002:31) at ?anon_0_ (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:129326:30) at throw (native) at resume (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:129380:30) at anonymous (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:129391:17) at tryCallOne (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:53:16) at anonymous (/private/var/folders/yw/6bx918xn4671rggfcdxz7fph0000gn/T/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:139:27) at apply (native) at anonymous (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:33521:26) at _callTimer (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:33440:17) at _callReactNativeMicrotasksPass (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:33470:17) at callReactNativeMicrotasks (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:33633:44) at __callReactNativeMicrotasks (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:6570:46) at anonymous (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:6382:45) at __guard (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:6554:15) at flushedQueue (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:6381:21) at callFunctionReturnFlushedQueue (http://192.168.100.7:19000/index.ts.bundle?platform=ios&dev=true&hot=false:6366:33)`

**Reply by dariocravero** [reply]
@elisalimli no, we don't use expo-router

**Comment by nandorojo** [comment]
Your import statement looks broken:

```ts
import { use query, gql } from "urql";
```

It should be `useQuery`, not `use query`.

**Reply by elisalimli** [reply]
Thanks for your feedback!

I imported it like `useQuery` it's just a typo made by Grammarly.
